app/modules.go

```diff
@@ -142,43 +142,25 @@ func appModules(
 	}
 }
 
-// orderBeginBlockers Tell the app's module manager how to set the order of
-// BeginBlockers, which are run at the beginning of every block.
-func orderBeginBlockers() []string {
-	return []string{
-		// Upgrades should be run VERY first
-		upgradetypes.ModuleName,
-		// Note: epochs' begin should be "real" start of epochs, we keep epochs beginblock at the beginning
-		epochstypes.ModuleName,
-		capabilitytypes.ModuleName,
-		minttypes.ModuleName,
-		poolincentivestypes.ModuleName,
-		distrtypes.ModuleName,
-		slashingtypes.ModuleName,
-		evidencetypes.ModuleName,
-		stakingtypes.ModuleName,
-		ibchost.ModuleName,
-		ibctransfertypes.ModuleName,
-		icatypes.ModuleName,
-		authtypes.ModuleName,
-		banktypes.ModuleName,
-		govtypes.ModuleName,
-		crisistypes.ModuleName,
-		genutiltypes.ModuleName,
-		authz.ModuleName,
-		paramstypes.ModuleName,
-		vestingtypes.ModuleName,
-		gammtypes.ModuleName,
-		incentivestypes.ModuleName,
-		lockuptypes.ModuleName,
-		poolincentivestypes.ModuleName,
-		tokenfactorytypes.ModuleName,
-		// superfluid must come after distribution and epochs
-		superfluidtypes.ModuleName,
-		bech32ibctypes.ModuleName,
-		txfeestypes.ModuleName,
-		wasm.ModuleName,
-	}
+// orderBeginBlockers returns the order of BeginBlockers, by module name.
+func orderBeginBlockers(allModuleNames []string) []string {
+	ord := partialord.NewPartialOrdering(allModuleNames)
+	// Upgrades should be run VERY first
+	// Epochs is set to be next right now, this in principle could change to come later / be at the end.
+	// But would have to be a holistic change with other pipelines taken into account.
+	ord.FirstElements(upgradetypes.ModuleName, epochstypes.ModuleName, capabilitytypes.ModuleName)
+
+	// Staking ordering
+	// TODO: Perhaps this can be relaxed, left to future work to analyze.
+	ord.Sequence(distrtypes.ModuleName, slashingtypes.ModuleName, evidencetypes.ModuleName, stakingtypes.ModuleName)
```

Comment:
Out of curiosity, what is the purpose of creating/using `Sequence` here? i.e. in what context would there be elements interspersed for `BeginBlock` logic? Is `BeginBlock` ordering being changed/affected anywhere else in the codebase?

Reply:
Is the part that's "just found" random, or is the ordering somehow externally controlled e.g. by validators?

Reply:
No, its deterministically found by solving for a total ordering that satisfies all constraints. If your interested in the details of the implemented algorithm, checkout the partialord package! (But TL;DR topological sort)